gno/r/launchpad_grc20/sale_grc20.gno

```diff
@@ -0,0 +1,266 @@
+package launchpad_grc20
+
+import (
+	"std"
+	"time"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/ufmt"
+)
+
+type Sale struct {
+	token          *Token
+	startTimestamp int64
+	endTimestamp   int64
+	pricePerToken  uint64
+	alreadySold    uint64
+	limitPerAddr   uint64
+	minGoal        uint64
+	maxGoal        uint64
+	owner          std.Address
+	buyers         *avl.Tree // address -> amount
+	finalized      bool
+	merkleRoot     string
+}
+
+var (
+	sales      *avl.Tree // sale ID -> sale
+	nextSaleID uint64
+)
+
+func init() {
+	sales = avl.NewTree()
+	nextSaleID = 1
+}
+
+func NewSale(tokenName, merkleRoot string, startTimestamp, endTimestamp int64, pricePerToken, limitPerAddr, minGoal, maxGoal uint64, mintToken bool) uint64 {
+	token := mustGetToken(tokenName)
+	token.admin.AssertCallerIsOwner()
+
+	if mintToken && !token.allowMint {
+		panic("token is not mintable")
+	}
+
+	owner := std.PrevRealm().Addr()
+
+	// Subtract 10 seconds to make it easier to create a sale that starts immediately (for testing purposes)
+	now := time.Now().Unix()
+	if startTimestamp < now-10 {
+		panic("start timestamp must be in the future")
+	}
+
+	if startTimestamp >= endTimestamp {
+		panic("invalid timestamps, start must be before end")
+	}
+
+	if minGoal > maxGoal {
+		panic("min goal must be less than max goal")
+	}
+
+	if pricePerToken == 0 {
+		panic("price per token must be greater than 0")
+	}
+
+	realmAddr := std.CurrentRealm().Addr()
+	if mintToken {
+		token.banker.Mint(realmAddr, maxGoal)
+	} else {
+		err := token.banker.Transfer(owner, realmAddr, maxGoal)
+		if err != nil {
+			panic("error while transferring tokens to the realm, " + err.Error())
+		}
+	}
+
+	sale := Sale{
+		token:          token,
+		startTimestamp: startTimestamp,
+		endTimestamp:   endTimestamp,
+		pricePerToken:  pricePerToken,
+		limitPerAddr:   limitPerAddr,
+		minGoal:        minGoal,
+		maxGoal:        maxGoal,
+		owner:          owner,
+		buyers:         avl.NewTree(),
+		finalized:      false,
+		merkleRoot:     merkleRoot,
+	}
+
+	saleID := nextSaleID
```

Comment:
it's generally recommended to use `p/demo/seqid` for auto-increment integer ids used as avl keys

the benefit are that the order is preserved while traversing the tree, which is useful for pagination

also if the id is used as a part of a composite key, `seqid` ensures it's length will be consistent and thus can be used as prefix

Reply:
resolved in [e11eae9](https://github.com/TERITORI/teritori-dapp/pull/1263/commits/e11eae9ddb2506138e618afd3f79ca7631bada2a)